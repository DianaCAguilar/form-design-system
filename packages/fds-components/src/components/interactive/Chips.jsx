import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import DenyIcon from 'fds-icons/lib/react/DenyIcon';

/**
 * Common pattern where we use href to generate whether
 * an href or a button is returned, and also allow a master
 * override where you can specify any element / component as
 * the root
 */
const baseElement = ({ href, rootEl }) => {
  let Element;
  if (rootEl) {
    Element = rootEl;
  } else if (href) {
    Element = 'a';
  } else {
    Element = 'button';
  }
  return Element;
};

const Chip = ({
  size,
  rootEl,
  theme,
  /* className is ignored */ className,
  isActive,
  children,
  label,
  subtitle,
  hasClose,
  ...rest
}) => {
  const Element = baseElement({ href: rest.href, rootEl });

  const rootClass = cx('fdsChip', {
    'fdsChip--active': isActive,
    'fdsChip--small': size === 'small',
    'fdsChip--large': size === 'large',
    'fdsChip--blue': theme === 'blue',
    'fdsChip--outline': theme === 'outline',
  });
  return (
    <Element className={rootClass} {...rest}>
      {label}
      {subtitle && <span className="fdsChip-subtitle">{subtitle}</span>}
      {hasClose && (
        <span className="fdsChip-close">
          <DenyIcon size="xs" />
        </span>
      )}
    </Element>
  );
};

Chip.propTypes = {
  label: PropTypes.string,
  /**
   * This allows for overriding of the Chip root element.
   * Meant to accomodate  `<Link />`
   */
  rootEl: PropTypes.func,
  isActive: PropTypes.bool,
  subtitle: PropTypes.string,
  className: PropTypes.string,
  children: PropTypes.any,
  hasClose: PropTypes.bool,
  value: PropTypes.string,
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  theme: PropTypes.oneOf(['blue', 'gray', 'outline']),
};

const Chips = ({ list, /* className is ignored */ className, ...topRest }) => (
  <ul className="fdsChip-ul">
    {list.map((listItem) => (
      <li className="fdsChip-li" key={listItem.value || listItem.label}>
        <Chip {...topRest} {...listItem} />
      </li>
    ))}
  </ul>
);

Chips.defaultProps = {
  theme: 'gray',
  size: 'medium',
};

Chips.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      /**
       * This allows for overriding of the Chip root element.
       * Meant to accomodate  `<Link />`
       */
      rootEl: PropTypes.func,
      isActive: PropTypes.bool,
      subtitle: PropTypes.string,
    })
  ),
  /** className will be ignored */
  className: PropTypes.string,
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  theme: PropTypes.oneOf(['blue', 'gray', 'outline']),
  /**
   * You can specify hasClose on all chips here. Feel free
   * to individually specify on each chip also
   */
  hasClose: PropTypes.bool,
};

export default Chips;
