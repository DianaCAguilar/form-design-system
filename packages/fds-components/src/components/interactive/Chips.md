```js
import Chips from './Chips';

const chips = [
  { label: 'Lorem', href: "#lorem" },
  { label: 'Ipsum', href: "#ipsum", isActive: true, subtitle: 'active' },
  { label: 'Dolor', href: "#dolor", hasClose: true },
  { label: 'Sit', onClick: () => console.log('more custom logic here'), isActive: false },
];

// Yes, this even works for Link components
const Link = ({ to, children, ...rest }) => (
  <a href={to} {...rest}>{children}</a>
)

const chipsTwo = [
  { label: 'Lorem Ipsum' },
  { label: 'Dolor Sit' },
  { rootEl: Link, label: 'hey', to: '#google' },
  { label: 'Sit Amet' },
  { label: 'Consectetur', onClick: () => console.log('more custom logic here') },
];

<div>
  <Chips list={chips} size="small" />
  <br />
  <Chips list={chips} theme="blue" size="small" />
  <br /><br />
  <Chips list={chips} />
  <br />
  <Chips list={chips} theme="blue"  />
  <br /><br />
  <Chips list={chipsTwo} hasClose={true} theme="outline"  />
  <br />
</div>
```