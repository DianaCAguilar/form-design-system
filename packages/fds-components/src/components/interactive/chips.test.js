import React from 'react';
import { shallow } from 'enzyme';

import Chips from './Chips';

const chips = [
    { label: 'Lorem', href: "#lorem" },
    { label: 'Ipsum', href: "#ipsum", isActive: true, subtitle: 'active' },
    { label: 'Dolor', href: "#dolor", hasClose: true },
    { label: 'Sit', onClick: () => console.log('custom logic here'), isActive: false },
];

/**
 * NOTE: I have *purposely* have not extensively tested this, to force me and my 
 * manager to revisit and figure out a robust testing strategy for more
 * complex components like this.
 */

const renderComponent = (props) => shallow(<Chips {...props} />);

describe('Chips component', () => {

    it('matches snapshot (default props)', () => {
        const component = renderComponent({
            list: chips,
        });
        expect(component).toMatchSnapshot();
    });
});
