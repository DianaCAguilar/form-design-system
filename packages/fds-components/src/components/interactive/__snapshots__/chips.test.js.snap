// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chips component matches snapshot (default props) 1`] = `
<ul
  className="fdsChip-ul"
>
  <li
    className="fdsChip-li"
    key="Lorem"
  >
    <Chip
      href="#lorem"
      label="Lorem"
      size="medium"
      theme="gray"
    />
  </li>
  <li
    className="fdsChip-li"
    key="Ipsum"
  >
    <Chip
      href="#ipsum"
      isActive={true}
      label="Ipsum"
      size="medium"
      subtitle="active"
      theme="gray"
    />
  </li>
  <li
    className="fdsChip-li"
    key="Dolor"
  >
    <Chip
      hasClose={true}
      href="#dolor"
      label="Dolor"
      size="medium"
      theme="gray"
    />
  </li>
  <li
    className="fdsChip-li"
    key="Sit"
  >
    <Chip
      isActive={false}
      label="Sit"
      onClick={[Function]}
      size="medium"
      theme="gray"
    />
  </li>
</ul>
`;
